optimum/neuron/modeling_seq2seq.py

```diff
@@ -0,0 +1,912 @@
+        # beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=input_ids.device)
+        beam_scores_device = "cpu"
+        beam_scores = torch.zeros((batch_size, num_beams), dtype=torch.float, device=beam_scores_device)
+        beam_scores[:, 1:] = -1e9
+        beam_scores = beam_scores.view((batch_size * num_beams,))
+
+        while True:
+            # prepare model inputs
+            # From max_length-sized input_ids, select first
+            # cur_len - 1 values.
+            update_indices = torch.stack(
+                [torch.arange(input_ids.size(0)), torch.tensor(cur_len - 1).repeat(input_ids.size(0))], dim=-1
+            )
+            input_ids_ = input_ids[update_indices[:, 0], update_indices[:, 1], None]
+            model_inputs = self.prepare_inputs_for_generation(input_ids_, **model_kwargs)
+
+            next_token_scores, next_tokens, next_indices = self(
+                **model_inputs,
+                return_dict=True,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+                beam_scores=beam_scores,
+            )
+
+            # stateless
+            beam_outputs = beam_scorer.process(
+                input_ids.to("cpu")[:, :cur_len],
+                next_token_scores.to("cpu"),
+                next_tokens.to("cpu"),
+                next_indices.to("cpu"),
+                pad_token_id=pad_token_id,
+                eos_token_id=eos_token_id,
+                beam_indices=beam_indices,
+            )
+
+            beam_scores = beam_outputs["next_beam_scores"]
+            beam_next_tokens = beam_outputs["next_beam_tokens"]
+            beam_idx = beam_outputs["next_beam_indices"]
+
+            update_indices = torch.stack(
+                [torch.arange(batch_beam_size), torch.tensor(cur_len - 1).repeat(batch_beam_size)], dim=-1
+            )
+            update_indices_2 = torch.stack(
+                [torch.arange(batch_beam_size), torch.tensor(cur_len).repeat(batch_beam_size)], dim=-1
+            )
+            # First select beam_indices
+            device = input_ids.device
+            beam_idx_device = beam_idx.to(device=input_ids.device)
+            input_ids[:, :] = input_ids[beam_idx_device.long(), :]
+
+            # Then append new tokens
+            input_ids[update_indices_2[:, 0], update_indices_2[:, 1], None] = (
+                beam_next_tokens.unsqueeze(-1).to(device).to(torch.long)
+            )
+            input_ids = input_ids * 1  # Hack to materialize tensor
+
+            # update generated ids, model inputs, and length for next step
+            model_kwargs = self._update_model_kwargs_for_xla_generation(
+                model_kwargs,
+                batch_size=batch_beam_size,
+                is_encoder_decoder=self.config.is_encoder_decoder,
+                max_length=stopping_criteria.max_length,
```

Comment:
Isn't stopping_criteria a list ? And it might not contain a MaxLengthCriteria ...

Reply:
I'm not sure... I just did as what's in `NeuronGenerationMixin`:
https://github.com/huggingface/optimum-neuron/blob/50ec86f387f0220c2f890749d7b776169c34c7da/optimum/neuron/generation/utils.py#L591

Reply:
As I said, `NeuronGenerationMixin` AFAIK has not been fully tested. If not specified explicitly, the stopping criteria is a list deduced from the parameters passed to generate. So I assume that somewhere in the code it is passed explicitly ?

Reply:
If it's not passed, it's instantiate here with the generation config:
https://github.com/huggingface/optimum-neuron/blob/50ec86f387f0220c2f890749d7b776169c34c7da/optimum/neuron/generation/utils.py#L991